--- kexec-tools-1.101/Makefile.orig	2006-10-20 13:42:52.000000000 -0400
+++ kexec-tools-1.101/Makefile	2006-10-20 13:43:22.000000000 -0400
@@ -45,7 +45,7 @@
 MAN_PAGES:= kexec/kexec.8
 MAN_PAGES+= kdump/kdump.8
 BINARIES_i386:=  $(SBINDIR)/kexec $(PKGLIBDIR)/kexec_test $(SBINDIR)/kdump
-BINARIES_x86_64:=$(SBINDIR)/kexec $(PKGLIBDIR)/kexec_test
+BINARIES_x86_64:=  $(SBINDIR)/kexec $(PKGLIBDIR)/kexec_test $(SBINDIR)/kdump
 BINARIES:=$(SBINDIR)/kexec $(BINARIES_$(ARCH)) 
 
 TARGETS:=$(BINARIES) $(MAN_PAGES)
