From 3beef142bc003d9cf2e957c6a21e51d661f9b13e Mon Sep 17 00:00:00 2001
From: Pingfan Liu <piliu@redhat.com>
Date: Thu, 28 Nov 2019 21:23:00 +0800
Subject: [PATCH] makedumpfile: remove -lebl

Signed-off-by: Pingfan Liu <piliu@redhat.com>
---
 makedumpfile-1.6.6/Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/makedumpfile-1.6.6/Makefile b/makedumpfile-1.6.6/Makefile
index 1fdb628..df21b93 100644
--- a/makedumpfile-1.6.6/Makefile
+++ b/makedumpfile-1.6.6/Makefile
@@ -50,7 +50,7 @@ OBJ_PART=$(patsubst %.c,%.o,$(SRC_PART))
 SRC_ARCH = arch/arm.c arch/arm64.c arch/x86.c arch/x86_64.c arch/ia64.c arch/ppc64.c arch/s390x.c arch/ppc.c arch/sparc64.c
 OBJ_ARCH=$(patsubst %.c,%.o,$(SRC_ARCH))
 
-LIBS = -ldw -lbz2 -lebl -ldl -lelf -lz
+LIBS = -ldw -lbz2 -ldl -lelf -lz
 ifneq ($(LINKTYPE), dynamic)
 LIBS := -static $(LIBS)
 endif
-- 
2.20.1

