--- kexec-tools-testing-20070330/kexec/arch/i386/kexec-bzImage.c.orig	2007-03-30 00:34:36.000000000 -0400
+++ kexec-tools-testing-20070330/kexec/arch/i386/kexec-bzImage.c	2007-08-27 15:30:42.000000000 -0400
@@ -326,6 +326,9 @@
 #define OPT_APPEND	(OPT_ARCH_MAX+0)
 #define OPT_RAMDISK	(OPT_ARCH_MAX+1)
 #define OPT_REAL_MODE	(OPT_ARCH_MAX+2)
+#define OPT_ARGS_ELF    (OPT_ARCH_MAX+3)
+#define OPT_ARGS_LINUX  (OPT_ARCH_MAX+4)
+#define OPT_ARGS_NONE   (OPT_ARCH_MAX+5)
 	static const struct option options[] = {
 		KEXEC_ARCH_OPTIONS
 		{ "debug",		0, 0, OPT_DEBUG },
@@ -334,6 +337,9 @@
 		{ "initrd",		1, 0, OPT_RAMDISK },
 		{ "ramdisk",		1, 0, OPT_RAMDISK },
 		{ "real-mode",		0, 0, OPT_REAL_MODE },
+		{ "args-elf",           0, NULL, OPT_ARGS_ELF },
+		{ "args-linux",         0, NULL, OPT_ARGS_LINUX },
+		{ "args-none",          0, NULL, OPT_ARGS_NONE },
 		{ 0, 			0, 0, 0 },
 	};
 	static const char short_options[] = KEXEC_ARCH_OPT_STR "d";
@@ -368,6 +374,11 @@
 		case OPT_REAL_MODE:
 			real_mode_entry = 1;
 			break;
+		case OPT_ARGS_ELF:
+		case OPT_ARGS_LINUX:
+		case OPT_ARGS_NONE:
+			/* we can ignore these options */
+			break;
 		}
 	}
 	command_line_len = 0;
