--- kexec-tools-1.101/Makefile~orig	2006-02-02 11:34:50.000000000 +0100
+++ kexec-tools-1.101/Makefile	2006-02-02 11:35:00.000000000 +0100
@@ -44,7 +44,7 @@
 
 MAN_PAGES:= kexec/kexec.8
 BINARIES_i386:=  $(SBINDIR)/kexec $(PKGLIBDIR)/kexec_test $(SBINDIR)/kdump
-BINARIES_x86_64:=$(SBINDIR)/kexec $(PKGLIBDIR)/kexec_test
+BINARIES_x86_64:=$(SBINDIR)/kexec $(PKGLIBDIR)/kexec_test $(SBINDIR)/kdump
 BINARIES:=$(SBINDIR)/kexec $(BINARIES_$(ARCH)) 
 
 TARGETS:=$(BINARIES) $(MAN_PAGES)
