--- kexec-tools-2.0.0/makedumpfile-1.3.5/Makefile.orig	2009-11-11 08:44:51.000000000 +0800
+++ kexec-tools-2.0.0/makedumpfile-1.3.5/Makefile	2010-05-30 00:25:17.350869858 +0800
@@ -32,7 +32,7 @@
 	$(CC) $(CFLAGS_ARCH) -c -o ./$@ ./$(@:.o=.c) 
 
 makedumpfile: $(SRC) $(OBJ_ARCH)
-	$(CC) $(CFLAGS) $(OBJ_ARCH) -o $@ $< -static -ldw -lelf -lz
+	$(CC) $(CFLAGS) $(OBJ_ARCH) -o $@ $< -ldw -lelf -lz
 	echo .TH MAKEDUMPFILE 8 \"$(DATE)\" \"makedumpfile v$(VERSION)\" \"Linux System Administrator\'s Manual\" > temp.8
 	grep -v "^.TH MAKEDUMPFILE 8" makedumpfile.8 >> temp.8
 	mv temp.8 makedumpfile.8
