--- kexec-tools-2.0.6/makedumpfile-1.5.6/Makefile.orig
+++ kexec-tools-2.0.6/makedumpfile-1.5.6/Makefile
@@ -66,7 +66,7 @@
 CFLAGS += -DUSESNAPPY
 endif
 
-all: makedumpfile
+all: makedumpfile eppic_makedumpfile.so
 
 $(OBJ_PART): $(SRC_PART)
 	$(CC) $(CFLAGS) -c -o ./$@ $(VPATH)$(@:.o=.c)
@@ -87,7 +87,7 @@
 	gzip -c ./makedumpfile.conf.5 > ./makedumpfile.conf.5.gz
 
 eppic_makedumpfile.so: extension_eppic.c
-	$(CC) $(CFLAGS) -shared -rdynamic -o $@ extension_eppic.c -fPIC -leppic -ltinfo
+	$(CC) $(CFLAGS) -shared -rdynamic -o $@ extension_eppic.c -I../eppic/libeppic -fPIC ../eppic/libeppic/libeppic.a -ltinfo
 
 clean:
 	rm -f $(OBJ) $(OBJ_PART) $(OBJ_ARCH) makedumpfile makedumpfile.8.gz makedumpfile.conf.5.gz
