--- kexec-tools-1.101/kexec/arch/ppc64/kexec-ppc64.c.orig	2007-04-10 10:53:44.000000000 -0400
+++ kexec-tools-1.101/kexec/arch/ppc64/kexec-ppc64.c	2007-04-10 10:53:58.000000000 -0400
@@ -369,8 +369,8 @@
 			}
 			rmo_base = ((unsigned long long *)buf)[0];
 			rmo_top = rmo_base + ((unsigned long long *)buf)[1];
-			if (rmo_top > 0x30000000UL)
-				rmo_top = 0x30000000UL;
+			if (rmo_top > 0x07C00000UL)
+				rmo_top = 0x07C00000UL;
 
 			fclose(file);
 			closedir(cdir);
